File: typescript/src/Backtracking/CombinationSum/CombinationSum.ts
class CombinationSum {
  candidates: number[]
  target: number

  constructor(candidates: number[], target: number) {
    this.candidates = candidates
    this.target = target
  }

  exec(): number[][]{
    const LEN: number = this.candidates.length;
    const res: number[][] = []

    // candidate stores one potential result
    const candidate: number[] = []
    const dfs = (i: number, remain: number): void => {
      // fail condition
      if(i === LEN || remain < 0) return;
      // success condition
      if(remain === 0){
        res.push([...candidate])
      }

      // using current number & stay
      candidate.push(this.candidates[i])
      dfs(i, remain - this.candidates[i])

      // not using current number & move to next
      candidate.pop()
      dfs(i + 1, remain - this.candidates[i])
    }

    dfs(0, this.target)

    return res;
  }

  exec2(): number[][] {
    const res: number[][] = []
    const LEN: number = this.candidates.length;

    const cur: number[] = []
    const dfs = (i: number, sum: number): void => {
      // we have finsihed the iteration
      if(i >= LEN || sum > this.target) return;
      // we have found result
      if(sum === this.target){
        res.push([...cur])
        return;
      }

      // include i and repeat
      cur.push(this.candidates[i])
      dfs(i, sum + this.candidates[i])

      // not include i and continue
      cur.pop()
      dfs(i + 1, sum)
    }

    dfs(0, 0)

    return res;
  }
}

export default CombinationSum;

File: python/src/backtracking/combination_sum/combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        LEN = len(candidates)

        cur = []

        def exec(i, acc):
            if i >= LEN or acc > target:
                return
            if acc == target:
                res.append(cur.copy())
                return

            # take current value and stay the index
            cur.append(candidates[i])
            exec(i, acc + candidates[i])

            # not take current value and continue
            cur.pop()
            exec(i + 1, acc)

        exec(0, 0)

        return res
